modules/gmake/src/proj.lua

```diff
@@ -0,0 +1,638 @@
+--
+-- @param prj
+--  project to print target name for
+---
+function proj.targetName(prj)
+	if isProject(prj) then
+		local name = prj.name
+		local ext = '.exe'
+		wl('TARGET = $(TARGETDIR)/%s%s', name, ext)
+	else
+		--- TODO: Override target name
+	end
+end
+
+
+---
+-- Prints the CPP flags for the given project.
+--
+-- @param prj
+--  project to print CPP flags for
+---
+function proj.cppFlags(prj)
+	if isProject(prj) then
+		local toolset = prj.compiler
+		local flags = toolset.getCppFlags(prj)
+
+		if #flags > 0 then
+			wl('ALL_CPPFLAGS += $(CPPFLAGS) -MMD -MP %s $(DEFINES) $(INCLUDES)', table.concat(flags, ' '))
+		else
+			wl('ALL_CPPFLAGS += $(CPPFLAGS) -MMD -MP $(DEFINES) $(INCLUDES)')
+		end
+	else
+		local cfg = prj
+		local toolset = cfg.compiler
+		local flags = toolset.getCppFlags(cfg)
+	
+		if #flags > 0 then
+			wl('ALL_CPPFLAGS += %s', table.concat(flags, ' '))
+		end
+	end	
+end
+
+
+---
+-- Prints the C flags for the given project.
+--
+-- @param prj
+--  project to print C flags for
+---
+function proj.cFlags(prj)
+	if isProject(prj) then
+		local toolset = prj.compiler
+		local flags = toolset.getCFlags(prj)
+
+		if #flags > 0 then
+			wl('ALL_CFLAGS += $(CFLAGS) $(ALL_CPPFLAGS) %s', table.concat(flags, ' '))
+		else
+			wl('ALL_CFLAGS += $(CFLAGS) $(ALL_CPPFLAGS)')
+		end
+	else
+		local cfg = prj
+		local toolset = cfg.compiler
+		local flags = toolset.getCFlags(cfg)
+
+		if #flags > 0 then
+			wl('ALL_CFLAGS += %s', table.concat(flags, ' '))
+		end
+	end
+end
+
+
+---
+-- Prints the CXX flags for the given project.
+--
+-- @param prj
+--  project to print CXX flags for
+---
+function proj.cxxFlags(prj)
+	if isProject(prj) then
+		local toolset = prj.compiler
+		local flags = toolset.getCxxFlags(prj)
+	
+		if #flags > 0 then
+			wl('ALL_CXXFLAGS += $(CXXFLAGS) $(ALL_CPPFLAGS) %s', table.concat(flags, ' '))
+		else
+			wl('ALL_CXXFLAGS += $(CXXFLAGS) $(ALL_CPPFLAGS)')
+		end
+	else
+		local cfg = prj
+		local toolset = cfg.compiler
+		local flags = toolset.getCxxFlags(cfg)
+	
+		if #flags > 0 then
+			wl('ALL_CXXFLAGS += %s', table.concat(flags, ' '))
+		end
+	end
+end
+
+
+---
+-- Prints the linker flags for the given project.
+--
+-- @param prj
+--  project to print linker flags for
+---
+function proj.linkFlags(prj)
+	if isProject(prj) then
+		local toolset = prj.linker
+		local flags = toolset.getLinkerFlags(prj)
+
+		if #flags > 0 then
+			wl('ALL_LDFLAGS = $(LDFLAGS) %s', table.concat(flags, ' '))
+		else
+			wl('ALL_LDFLAGS = $(LDFLAGS)')
+		end
+	else
+		local cfg = prj
+		local toolset = cfg.linker
+		local flags = toolset.getLinkerFlags(cfg)
+
+		if #flags then
+			wl('ALL_LDFLAGS += %s', table.concat(flags, ' '))
+		end
+	end
+end
+
+
+---
+-- Prints all configurations for the project.
+--
+-- @param prj
+--  project to print configurations for
+---
+function proj.configurations(prj)
+	wl('# Configuration-level overrides')
+	for _, cfg in ipairs(prj.configs) do
```

Comment:
FYI, using `for i = 1, #configs` is measurably faster than `ipairs()`. I've been avoiding `ipairs()` in my own code.